File: frontend/src/components/technical-reports/diameterTotals.ts
export function sumDiameterColumns(
  rows: Array<Record<string, number>>,
  diameters: string[],
): Record<string, number> {
  const totals: Record<string, number> = {};
  for (const diameter of diameters) {
    let sum = 0;
    for (const row of rows) {
      sum += Number(row[diameter]) || 0;
    }
    totals[diameter] = sum;
  }
  return totals;
}

/** Sum all diameter values in a single row (horizontal total for OPER.) */
export function sumDiameterRow(
  row: Record<string, number>,
  diameters: string[],
): number {
  let sum = 0;
  for (const d of diameters) {
    sum += Number(row[d]) || 0;
  }
  return sum;
}

File: backend/core/technical_reports/diameter_totals.py
from __future__ import annotations

from typing import Any


def sum_diameter_columns(
    rows: list[dict[str, Any] | None],
    diameters: list[str],
) -> dict[str, int]:
    totals = {diameter: 0 for diameter in diameters}
    for row in rows:
        if not row:
            continue
        for diameter in diameters:
            value = row.get(diameter)
            if not value:
                continue
            try:
                totals[diameter] += int(value)
            except (TypeError, ValueError):
                continue
    return totals


def sum_diameter_row(
    row: dict[str, Any] | None,
    diameters: list[str],
) -> int:
    if not row:
        return 0
    total = 0
    for diameter in diameters:
        value = row.get(diameter)
        if not value:
            continue
        try:
            total += int(value)
        except (TypeError, ValueError):
            continue
    return total

